### ISSUE TITLE
Investigate out of the box JSON::Serializable support

### ISSUE THREAD

**Issue by Blacksmoke16** [amberframework/granite#250]
I would like to start a discussion on how doable support for Crystal `JSON::Serializable` would be for Granite.  

If this works out how I think it will this would be a major improvement on the ability for models to work much better with JSON APIs.  It would allow:

1.  Better `.from_json` support 
2.  Possibly allow for the `JSON::Field` annotations to better control serialization/deserialization.  
3.  Tap into the `after_initialize` event.  I..e you could run validations after consuming a JSON object to check if the provided values are allowed with the provided `validate_*` validators.  Or populate meta fields based on JSON values + a DB query for example.

I'm not too familiar with how/what would have to be done to support this (or if its even doable).  So i wanted to see what thoughts you guys have.

**Comment by robacarp** [comment]
@Blacksmoke16 I'm all for ideas! The current json integration is minimal at best and various ideas have been floated in the past, both here and on Amber.

**Comment by Blacksmoke16** [comment]
Support for this would be amazing.

I been looking into it here and there but im not too sure what is going on still.  

It seems `from_json` is trying to create a `Symbol` type that is not able to be instantiated.  

```
instantiating '(Symbol | Nil):Class#new(JSON::PullParser)'
in /usr/share/crystal/src/json/from_json.cr:196: expanding macro

  {% begin %}
  ^

in macro 'macro_140137072880928' /usr/share/crystal/src/json/from_json.cr:196, line 18:

   1. 
   2.     # Here we store types that are not primitive types
   3.     
   4. 
   5.     
   6.       
   7.         
   8.       
   9.     
  10.       
  11.         return pull.read_null if pull.kind == :null
  12.       
  13.     
  14. 
  15.     # If after traversing all the types we are left with just one
  16.     # non-primitive type, we can parse it directly (no need to use `read_raw`)
  17.     
> 18.       return Symbol.new(pull)
  19.     
  20.   

undefined method 'new' for Symbol:Class
```

**Comment by Blacksmoke16** [comment]
Fixed with #253
